crates/red_knot_workspace/src/site_packages.rs

```diff
@@ -8,15 +8,68 @@
 //! reasonably ask us to type-check code assuming that the code runs
 //! on Linux.)
 
+use std::fmt;
 use std::io;
+use std::num::NonZeroUsize;
+use std::ops::Deref;
 
 use ruff_db::system::{System, SystemPath, SystemPathBuf};
 
-/// Attempt to retrieve the `site-packages` directory
-/// associated with a given Python installation.
+/// The value given by the `home` key in `pyvenv.cfg` files.
+///
+/// This is equivalent to `{sys_prefix_path}/bin`, and points
+/// to a directory in which a Python executable can be found.
+/// Confusingly, it is *not* the same as the [`PYTHONHOME`]
+/// environment variable that Python provides! However, it's
+/// consistent among all mainstream creators of Python virtual
+/// environments (the stdlib Python `venv` module, the third-party
+/// `virtualenv` library, and `uv`), was specified by
+/// [the original PEP adding the `venv` module],
+/// and it's one of the few fields that's read by the Python
+/// standard library's `site.py` module.
+///
+/// Other values, such as the path to the Python executable or the
+/// base-executable `sys.prefix` value, are either only provided in
+/// `pyvenv.cfg` files by some virtual-environment creators,
+/// or are included under different keys depending on which
+/// virtual-environment creation tool you've used.
+///
+/// [`PYTHONHOME`]: https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHOME
+/// [the original PEP adding the `venv` module]: https://peps.python.org/pep-0405/
+#[derive(Debug, PartialEq, Eq)]
+struct PythonHomePath(SystemPathBuf);
```

Comment:
This might be a Python habit of you but I generally prefer to have the public API at the top of the module and move module internal structs and functions further down. I find that it helps reading the code because I'm mostly concerned about the public API and only want to dive into the internals if I have to make changes/fix something. 

I think it can also help reviewers because I'm now forced to first read all internals before I get to the point where I see the external API, which is probably the most interesting thing to review.

Reply:
I don't think this is a particular convention I'm following; I think I'm just not thinking too much about it, which is on me :-)

I'll restructure this module and try to think more about this in the future, thanks.